### PR TITLE
Transform dynamic watches integration tests into unit tests

### LINKED ISSUE #908: Improve dynamic watches integration tests
Many tests in https://github.com/elastic/cloud-on-k8s/blob/master/operators/pkg/controller/common/watches/handler_integration_test.go there rely on an anti-pattern: "wait until timeout is reached and check we didn't receive any event". I think it's OK to have one or two of these but not have most test rely on it.

We currently use this pattern in both `CheckReconcileNotCalledWithin` (using an arbitrary 1sec wait)  and `CheckReconcileCalledIn` (using `test.Timeout/N`).

This makes the tests quite long. Also if we want to increase the default 5sec test timeout to, for example, 30sec, these tests suddenly become way longer for no real reason.

We should find a way to test the watch handler behaviour differently.

### PR DESCRIPTION
**Pull request by sebgl**
This commit sets up a unit test to cover the integration test scenario
we had for the DynamicEnqueueRequest handler. A very basic integration
test is left to ensure everything correctly fits together, but most of
the logic we wanted to test (and a bit more) is moved into the unit test.

Main benefits:
* remove complexity from IT (edge cases of 1-2 potential reconciliations,
wait for something to not happen)
* remove dependency to test.Timeout in most tests (used both in
`CheckReconcileCalledIn` and `CheckReconcileNotCalledWithin`)
* faster execution time

Fixes #908.

### CODE REVIEW

File: operators/pkg/utils/test/integration.go [hunk 1]
```diff
@@ -41,3 +46,29 @@ func RunWithK8s(m *testing.M, crdPath string) {
 	t.Stop()
 	os.Exit(code)
 }
+
+// StartManager sets up a manager and controller to perform reconciliations in background.
+// It must be stopped by calling the returned function.
+func StartManager(t *testing.T, addToMgrFunc func(manager.Manager, operator.Parameters) error, parameters operator.Parameters) (k8s.Client, func()) {
```
Comment (sebgl) [inline operators/pkg/utils/test/integration.go]: This comes from https://github.com/elastic/cloud-on-k8s/pull/906, should disappear from the diff once #906 is merged.

**Comment by sebgl** [comment]
jenkins test this please
